networkx/algorithms/bipartite/EFM_algorithms.py

```diff
@@ -0,0 +1,291 @@
+    M: dict
+       dictionary that represents a maximum matching in G.
+       If M is none, the function will calculate a maximum matching.
+    top_nodes: list
+                if graph is not all connected top_nodes is the set of the top nodes in the bipartite graph G
+
+    Returns
+    -------
+    EFM: tuple of sets
+        The partition returns as a tuple of 4 sets of vertices:
+        X_L,X_S,Y_L,Y_S where X_L,Y_L are the "good vertices" of G and
+        X_S,Y_S are the "bad vertices" of G where no envy-free matching exists according to THM 1.3 in the article.
+
+    References
+    ----------
+    .. [1] "Envy-free Matchings in Bipartite Graphs and their Applications to Fair Division",
+    by Elad Aigner-Horev and Erel Segal-Halevi (2022), https://arxiv.org/abs/1901.09527
+    Algorithm 1: Finding the EFM partition of a bipartite graph.
+
+    Examples
+    --------
+        Example 1: Perfect matching
+        >>> Graph=nx.complete_bipartite_graph(3,3)
+        >>> Matching=nx.bipartite.hopcroft_karp_matching(Graph)
+        >>> _EFM_partition(Graph,Matching)
+        ({0, 1, 2}, set(), {3, 4, 5}, set())
+
+        Where there exists a perfect matching the maximum envy free matching is the perfect matching.
+
+        Example 2: Non-empty envy-free matching
+        >>> Graph=nx.Graph([(0,3),(3,0),(0,4),(4,0),(1,4),(4,1),(2,4),(4,2)])
+        >>> Matching={0:3,3:0,1:4,4:1}
+        >>> _EFM_partition(Graph,Matching)
+        ({0}, {1, 2}, {3}, {4})
+
+        Here the graph contains non-empty envy-free matching so X_L,Y_L are not empty.
+
+        Example 3: Odd path
+        >>> Graph=nx.Graph([(0,3),(3,0),(1,3),(3,1),(1,4),(4,1),(2,4),(4,2)])
+        >>> Matching={0:3,3:0,4:1,1:4}
+        >>> _EFM_partition(Graph,Matching)
+        (set(), {0, 1, 2}, set(), {3, 4})
+
+        Like presented in the article, odd path contains an empty envy-free matching so X_L and Y_L are empty in the partition.
+
+        Example 4: Y-path-saturated graph
+        >>> Graph=nx.Graph([(0,6),(6,0),(1,6),(6,1),(1,7),(7,1),(2,6),(6,2),(2,8),(8,2),(3,9),(9,3),(3,6),(6,3),(4,8),(8,4),(4,7),(7,4),(5,9),(9,5)])
+        >>> Matching={0:6,6:0,1:7,7:1,2:8,8:2,3:9,9:3}
+        >>> _EFM_partition(Graph,Matching)
+        (set(), {0, 1, 2, 3, 4, 5}, set(), {8, 9, 6, 7})
+
+        Like presented in the article, Y-path-saturated graph contains an empty envy-free matching so X_L and Y_L are empty in the partition.
+
+    """
+    if M is None:
+        M = nx.bipartite.maximum_matching(G, top_nodes=top_nodes)
+    X, Y = nx.bipartite.sets(G, top_nodes=top_nodes)
+    X_subsets, Y_subsets = __M_alternating_sequence__(G, M, top_nodes)
+    X_S = set()
+    for subset in X_subsets:
+        X_S.update(subset)
+
+    Y_S = set()
+    for subset in Y_subsets:
+        Y_S.update(subset)
+
+    return set(X) - X_S, X_S, set(Y) - Y_S, Y_S
+
+
+@not_implemented_for("directed", "multigraph")
+def envy_free_matching(G, top_nodes=None):
```

Comment:
The name of this function should indicate that it returns an envy-free matching of maximum cardinality. Perhaps "largest_envy_free_matching" or "max_size_envy_free_matching".